From 4de7169f97c5e320ee48432aacd09e211443cf33 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:31 -0400
Subject: [fs] sunrpc: Allow svc_reserve() to notify TCP socket that space has been freed

Message-id: <1410898849-22126-88-git-send-email-bfields@redhat.com>
Patchwork-id: 95732
O-Subject: [Patch RHEL7 087/165] SUNRPC: Allow svc_reserve() to notify TCP socket that space has been freed
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 518776800c094a518ae6d303660b57f1400eb1eb
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/sunrpc/svc_xprt.h b/include/linux/sunrpc/svc_xprt.h
index 5d9d6f8..ce6e418 100644
--- a/include/linux/sunrpc/svc_xprt.h
+++ b/include/linux/sunrpc/svc_xprt.h
@@ -25,6 +25,7 @@ struct svc_xprt_ops {
  void  (*xpo_detach)(struct svc_xprt *);
  void  (*xpo_free)(struct svc_xprt *);
  int  (*xpo_secure_port)(struct svc_rqst *);
+ void  (*xpo_adjust_wspace)(struct svc_xprt *);
 };
 
 struct svc_xprt_class {
diff --git a/net/sunrpc/svc_xprt.c b/net/sunrpc/svc_xprt.c
index 9cfa391..6666c67 100644
--- a/net/sunrpc/svc_xprt.c
+++ b/net/sunrpc/svc_xprt.c
@@ -448,6 +448,8 @@ void svc_reserve(struct svc_rqst *rqstp, int space)
   atomic_sub((rqstp->rq_reserved - space), &xprt->xpt_reserved);
   rqstp->rq_reserved = space;
 
+  if (xprt->xpt_ops->xpo_adjust_wspace)
+   xprt->xpt_ops->xpo_adjust_wspace(xprt);
   svc_xprt_enqueue(xprt);
  }
 }
diff --git a/net/sunrpc/svcsock.c b/net/sunrpc/svcsock.c
index cd530b8..83ddfc5 100644
--- a/net/sunrpc/svcsock.c
+++ b/net/sunrpc/svcsock.c
@@ -475,6 +475,14 @@ static void svc_tcp_write_space(struct sock *sk)
  svc_write_space(sk);
 }
 
+static void svc_tcp_adjust_wspace(struct svc_xprt *xprt)
+{
+ struct svc_sock *svsk = container_of(xprt, struct svc_sock, sk_xprt);
+
+ if (svc_tcp_has_wspace(xprt))
+  clear_bit(SOCK_NOSPACE, &svsk->sk_sock->flags);
+}
+
 /*
  * See net/ipv6/ip_sockglue.c : ip_cmsg_recv_pktinfo
  */
@@ -1289,6 +1297,7 @@ static struct svc_xprt_ops svc_tcp_ops = {
  .xpo_has_wspace = svc_tcp_has_wspace,
  .xpo_accept = svc_tcp_accept,
  .xpo_secure_port = svc_sock_secure_port,
+ .xpo_adjust_wspace = svc_tcp_adjust_wspace,
 };
 
 static struct svc_xprt_class svc_tcp_class = {
-- 
1.7.1